File: apps/api/src/auto-healing-autonomy.ts
import {
  listAutoHealingAutonomyContexts,
  queryVerifiedAutoHealingRecoveries,
  type AutoHealingAutonomyContext,
  type AutoHealingRun,
} from "@janusly/data";
import {
  WorkflowSchema,
  classifyTechnicalRecoveryRepair,
  evaluateTechnicalRecoveryAutonomy,
  parseValidationEvidenceLevel,
  type TechnicalRecoveryAutonomyAssessment,
} from "@janusly/shared";

export type AutoHealingRunWithAutonomy = AutoHealingRun & {
  autonomyAssessment: TechnicalRecoveryAutonomyAssessment;
};

export async function assessAutoHealingAutonomyRows(
  orgId: string,
  rows: readonly AutoHealingRun[],
): Promise<AutoHealingRunWithAutonomy[]> {
  if (rows.length === 0) return [];
  const [contexts, counts] = await Promise.all([
    listAutoHealingAutonomyContexts(
      orgId,
      rows.map((row) => row.deadLetterId),
    ),
    queryVerifiedAutoHealingRecoveries(
      orgId,
      rows.map((row) => row.signature),
    ),
  ]);
  const contextById = new Map(
    contexts.map((context) => [context.deadLetterId, context]),
  );
  const countBySignature = new Map(
    counts.map((item) => [item.signature, item.count]),
  );
  return rows.map((row) => ({
    ...row,
    autonomyAssessment: buildAutoHealingAutonomyAssessment({
      row,
      context: contextById.get(row.deadLetterId) ?? null,
      priorVerifiedRecoveries: countBySignature.get(row.signature) ?? 0,
    }),
  }));
}

export async function assessAutoHealingAutonomyRow(
  row: AutoHealingRun,
  context?: AutoHealingAutonomyContext,
): Promise<TechnicalRecoveryAutonomyAssessment> {
  const [resolvedContext, counts] = await Promise.all([
    context
      ? Promise.resolve(context)
      : listAutoHealingAutonomyContexts(
          row.orgId,
          [row.deadLetterId],
        ).then((items) => items[0] ?? null),
    queryVerifiedAutoHealingRecoveries(row.orgId, [row.signature]),
  ]);
  return buildAutoHealingAutonomyAssessment({
    row,
    context: resolvedContext,
    priorVerifiedRecoveries: counts[0]?.count ?? 0,
  });
}

export function buildAutoHealingAutonomyAssessment(input: {
  row: AutoHealingRun;
  context: AutoHealingAutonomyContext | null;
  priorVerifiedRecoveries: number;
}): TechnicalRecoveryAutonomyAssessment {
  const original = WorkflowSchema.safeParse(input.context?.workflowJson);
  const candidate = WorkflowSchema.safeParse(input.row.proposedPatchJson);
  const failure =
    readErrorCode(input.context?.errorJson) === "worker_stalled"
      ? "stalled_node"
      : "terminal_node_failure";
  const repairClass =
    original.success && candidate.success && input.context
      ? classifyTechnicalRecoveryRepair({
          original: original.data,
          candidate: candidate.data,
          failingNodeId: input.context.nodeId,
        })
      : null;

  return evaluateTechnicalRecoveryAutonomy({
    contract: original.success
      ? original.data.recovery?.contract
      : null,
    failure,
    repairClass,
    validationEvidenceLevel: parseValidationEvidenceLevel(
      input.row.validationEvidenceLevel,
    ),
    priorVerifiedRecoveries: input.priorVerifiedRecoveries,
    affectedExecutions: 1,
    rollbackReady: original.success && candidate.success,
  });
}

function readErrorCode(value: unknown): string | null {
  if (value == null || typeof value !== "object" || Array.isArray(value)) {
    return null;
  }
  return typeof (value as Record<string, unknown>).code === "string"
    ? (value as Record<string, string>).code
    : null;
}

File: internal/httpapi/autohealingautonomy.go
package httpapi

import (
	"context"
	"encoding/json"

	"github.com/johnny4young/janusly/internal/domain"
	"github.com/johnny4young/janusly/internal/store"
)

const autoHealingAutonomyFactLimit = 200

// assessAutoHealingRows resolves all tenant-scoped facts in one bounded query
// and derives the same fail-closed authority explanation as the Node baseline.
// The projection explains policy only; decide/apply remains operator-owned.
func assessAutoHealingRows(ctx context.Context, q *store.Queries, orgID string, rows []store.AutoHealingRun) ([]domain.TechnicalRecoveryAutonomyAssessment, error) {
	if len(rows) == 0 {
		return []domain.TechnicalRecoveryAutonomyAssessment{}, nil
	}
	ids := make([]string, 0, min(len(rows), autoHealingAutonomyFactLimit))
	seen := make(map[string]struct{}, cap(ids))
	for _, row := range rows {
		if row.ID == "" || len(ids) == autoHealingAutonomyFactLimit {
			continue
		}
		if _, duplicate := seen[row.ID]; duplicate {
			continue
		}
		seen[row.ID] = struct{}{}
		ids = append(ids, row.ID)
	}
	facts, err := q.ListAutoHealingAutonomyFacts(ctx, store.ListAutoHealingAutonomyFactsParams{
		OrgID: orgID, Ids: ids,
	})
	if err != nil {
		return nil, err
	}
	factByID := make(map[string]store.ListAutoHealingAutonomyFactsRow, len(facts))
	for _, fact := range facts {
		factByID[fact.AutoHealingID] = fact
	}

	assessments := make([]domain.TechnicalRecoveryAutonomyAssessment, 0, len(rows))
	for _, row := range rows {
		fact := factByID[row.ID]
		var original *domain.TechnicalRecoveryWorkflow
		if fact.ContextFound {
			original, _ = domain.ParseTechnicalRecoveryWorkflow(fact.WorkflowJson)
		}
		candidate, candidateOK := domain.ParseTechnicalRecoveryWorkflow(row.ProposedPatchJson)
		originalOK := original != nil
		var contract *domain.RecoveryContract
		if originalOK && original.Recovery != nil {
			contract = original.Recovery.Contract
		}
		repairClass := ""
		if originalOK && candidateOK {
			repairClass = domain.ClassifyTechnicalRecoveryRepair(original, candidate, fact.NodeID)
		}
		failure := domain.TechnicalFailureTerminal
		if fact.ContextFound && autoHealingErrorCode(fact.ErrorJson) == "worker_stalled" {
			failure = domain.TechnicalFailureStalled
		}
		priorVerified := int(max(fact.PriorVerifiedRecoveries, 0))
		evidence := ""
		if row.ValidationEvidenceLevel.Valid {
			evidence = row.ValidationEvidenceLevel.String
		}
		assessments = append(assessments, domain.EvaluateTechnicalRecoveryAutonomy(domain.TechnicalRecoveryAutonomyInput{
			Contract: contract, Failure: failure, RepairClass: repairClass,
			ValidationEvidenceLevel: evidence, PriorVerifiedRecoveries: priorVerified,
			AffectedExecutions: 1, RollbackReady: originalOK && candidateOK,
		}))
	}
	return assessments, nil
}

func autoHealingErrorCode(raw json.RawMessage) string {
	var payload struct {
		Code string `json:"code"`
	}
	if json.Unmarshal(raw, &payload) != nil {
		return ""
	}
	return payload.Code
}
